js/render/draw_background.js

```diff
@@ -70,10 +58,48 @@ function drawBackground(painter, source, layer) {
     // we don't have so much going on in the stencil buffer.
     var coords = pyramid.coveringTiles(transform);
     for (var c = 0; c < coords.length; c++) {
-        gl.setPosMatrix(painter.calculatePosMatrix(coords[c]));
+        var coord = coords[c];
+        var tileSize = 512;
+        if (image && source) {
+            var imageSizeScaledA = [
+                (imagePosA.size[0] * image.fromScale),
+                (imagePosA.size[1] * image.fromScale)
+            ];
+            var imageSizeScaledB = [
+                (imagePosB.size[0] * image.toScale),
+                (imagePosB.size[1] * image.toScale)
+            ];
+
+            gl.uniform2fv(shader.u_patternscale_a, [
+                1 / pixelsToTileUnits(coord, imageSizeScaledA[0], painter.transform.tileZoom),
+                1 / pixelsToTileUnits(coord, imageSizeScaledA[1], painter.transform.tileZoom)
+            ]);
+
+            gl.uniform2fv(shader.u_patternscale_b, [
+                1 / pixelsToTileUnits(coord, imageSizeScaledB[0], painter.transform.tileZoom),
+                1 / pixelsToTileUnits(coord, imageSizeScaledB[1], painter.transform.tileZoom)
+            ]);
+            var tileSizeAtNearestZoom = tileSize * Math.pow(2, painter.transform.tileZoom - coord.z);
+
+            var offsetAx = ((tileSizeAtNearestZoom / imageSizeScaledA[0]) % 1) * (coord.x + coord.w * Math.pow(2, coord.z));
+            var offsetAy = ((tileSizeAtNearestZoom / imageSizeScaledA[1]) % 1) * coord.y;
+
+            var offsetBx = ((tileSizeAtNearestZoom / imageSizeScaledB[0]) % 1) * (coord.x + coord.w * Math.pow(2, coord.z));
+            var offsetBy = ((tileSizeAtNearestZoom / imageSizeScaledB[1]) % 1) * coord.y;
+
+            gl.uniform2fv(shader.u_offset_a, [offsetAx, offsetAy]);
+            gl.uniform2fv(shader.u_offset_b, [offsetBx, offsetBy]);
+        }
+
+        gl.setPosMatrix(painter.calculatePosMatrix(coord));
         gl.drawArrays(gl.TRIANGLE_STRIP, 0, painter.tileExtentBuffer.itemCount);
     }
 
     gl.stencilMask(0x00);
     gl.stencilFunc(gl.EQUAL, 0x80, 0x80);
 }
+
+function pixelsToTileUnits(coord, pixelValue, z) {
+    return pixelValue * (extent / (512 * Math.pow(2, z - coord.z)));
+}
```

Comment:
Can you find a way to eliminate the code duplication between :point_up_2: and [`Tile`](https://github.com/mapbox/mapbox-gl-js/blob/master/js/source/tile.js#L43)?